src/TreeSitter/Importing.hs

```diff
@@ -128,7 +158,56 @@ instance Monad m => Monad (MaybeC m) where
     a' <- a
     case a' of
       Nothing -> pure Nothing
-      Just a -> runMaybeC $ f a
+      Just a  -> runMaybeC $ f a
 
 instance MonadIO m => MonadIO (MaybeC m) where
   liftIO = MaybeC . fmap Just . liftIO
+
+
+newtype FieldName = FieldName { getFieldName :: String }
+  deriving (Eq, Ord, Show)
+
+
+-- | Generic construction of ASTs from a 'Map.Map' of named fields.
+--
+--   Product types (specifically, record types) are constructed by looking up the node for each corresponding field name in the map, moving the cursor to it, and then invoking 'buildNode' to construct the value for that field. Leaf types are constructed as a special case of product types.
+--
+--   Sum types are constructed by attempting to build each constructor nondeterministically. This should instead use the current node’s symbol to select the corresponding constructor deterministically.
+class GBuilding f where
+  gbuildNode :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => Map.Map FieldName Node -> m (f a)
+
+-- Build an empty node of given type
+  gbuildEmpty :: Alternative m => m (f a)
+
+instance (GBuilding f, GBuilding g) => GBuilding (f :*: g) where
+  gbuildNode fields = (:*:) <$> gbuildNode @f fields <*> gbuildNode @g fields
+  gbuildEmpty = (:*:) <$> gbuildEmpty @f <*> gbuildEmpty @g
+
+instance (GBuilding f, GBuilding g) => GBuilding (f :+: g) where
+  gbuildNode fields = L1 <$> gbuildNode @f fields <|> R1 <$> gbuildNode @g fields
```

Comment:
We will eventually wanna lose the `Alternation` here in favor of committed choice.